domain/state.go

```diff
@@ -44,3 +50,31 @@ func (st *StateBase) DB() (database.TxnRunner, error) {
 
 	return st.db, nil
 }
+
+// Prepare prepares a SQLair query. If the query has been prepared previously it
+// is retrieved from the statement cache.
+//
+// Note that because the type samples are not considered when retrieving a query
+// from the cache, it is possible that two queries may have identical text, but
+// use different types. Retrieving the wrong query would result in an error when
+// the query was passed the wrong type at execution.
+//
+// The likelihood of this happening is low since the statement cache is scoped
+// to individual domains meaning that the two identically worded statements
+// would have to be in the same state package. This issue should be relatively
+// rare and caught by QA if present.
+func (st *StateBase) Prepare(query string, typeSamples ...any) (*sqlair.Statement, error) {
+	st.stmtMutex.Lock()
+	defer st.stmtMutex.Unlock()
+
+	if stmt, ok := st.stmts[query]; ok && stmt != nil {
+		return stmt, nil
+	}
+
+	stmt, err := sqlair.Prepare(query, typeSamples...)
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing:")
+	}
+	st.stmts[query] = stmt
+	return stmt, nil
+}
```

Comment:
```suggestion
func (st *StateBase) Prepare(query string, typeSamples ...any) (*sqlair.Statement, error) {
	// Take a read lock to check if the statement is already prepared.
	st.stmtMutex.RLock()
	if stmt, ok := st.stmts[query]; ok && stmt != nil {
		st.stmtMutex.RUnlock()
		return stmt, nil
	}
	st.stmtMutex.RUnlock()

	// Grab the write lock to prepare the statement.
	st.stmtMutex.Lock()
	defer st.stmtMutex.Unlock()

	stmt, err := sqlair.Prepare(query, typeSamples...)
	if err != nil {
		return nil, errors.Annotate(err, "preparing:")
	}
	st.stmts[query] = stmt
	return stmt, nil
}
```

Reply:
Ah thanks, had meant to change this, forgot I hadn't